UniversalAlgebra/Terms/Setoid.lagda

```diff
@@ -74,48 +72,55 @@ module Environment (M : SetoidAlgebra α ℓ) where
 
  open SetoidAlgebra M
 
- open IsEquivalence renaming ( refl  to  reflE
-                             ; sym   to  symE
-                             ; trans to  transE )
-
  open Setoid        renaming ( refl  to  reflS
                              ; sym   to  symS
                              ; trans to  transS)
 
  -- Equality in M's interpretation of its sort.
- _≃_ : Den .Carrier → Den .Carrier → Type ℓ
- _≃_ = Den ._≈_
+ _≃_ : Domain .Carrier → Domain .Carrier → Type ℓ
+ _≃_ = Domain ._≈_
 
 
  -- An environment for Γ maps each variable `x : Γ` to an element of M.
  -- Equality of environments is defined pointwise.
  Env : Type χ → Setoid _ _
- Env Γ .Carrier                     = (x : Γ) → Den .Carrier
- Env Γ ._≈_ ρ ρ'                    = (x : Γ) → ρ x ≃ ρ' x
- Env Γ .isEquivalence .reflE      x = reflS  Den
- Env Γ .isEquivalence .symE     h x = symS   Den (h x)
- Env Γ .isEquivalence .transE g h x = transS Den (g x) (h x)
+ Env Γ = record { Carrier = (x : Γ) → Carrier Domain
+
+                ; _≈_ = λ ρ ρ' → (x : Γ) → ρ x ≃ ρ' x
+
+                ; isEquivalence =
+                   record { refl = λ _ → reflS Domain
+                          ; sym = λ h x → symS Domain (h x)
+                          ; trans = λ g h x → transS Domain (g x) (h x)
+                          }
+                }
+
 
 
  -- Interpretation of terms is iteration on the W-type.
  -- The standard library offers `iter' (on sets), but we need this to be a Func (on setoids).
- ⦅_⦆ : (t : Term Γ) → Func (Env Γ) Den
+ open Func renaming ( f to apply )
+
+ ⦅_⦆ : (t : Term Γ) → Func (Env Γ) Domain
  apply ⦅ ℊ x ⦆         ρ      =  ρ x
  cong  ⦅ ℊ x ⦆         ρ₁=ρ₂  =  ρ₁=ρ₂ x
- apply ⦅ node f args ⦆  ρ      =  apply den (f , λ i → apply ⦅ args i ⦆ ρ)
- cong  ⦅ node f args ⦆  ρ₁=ρ₂  =  cong den (refl , λ i → cong ⦅ args i ⦆ ρ₁=ρ₂)
+ apply ⦅ node f args ⦆  ρ      =  apply Interp (f , λ i → apply ⦅ args i ⦆ ρ)
+ cong  ⦅ node f args ⦆  ρ₁=ρ₂  =  cong Interp (refl , λ i → cong ⦅ args i ⦆ ρ₁=ρ₂)
 
 
  -- An equality between two terms holds in a model
  -- if the two terms are equal under all valuations of their free variables.
  Equal : ∀ {Γ : Type χ} (p q : Term Γ) → Type _
  Equal p q = ∀ (ρ : Env _ .Carrier) →  apply ⦅ p ⦆ ρ ≃ apply ⦅ q ⦆ ρ
 
+
  -- Equal is an equivalence relation.
  isEquiv : IsEquivalence (Equal {Γ = Γ})
- reflE  isEquiv     ρ = reflS  Den
- symE   isEquiv   i ρ = symS   Den (i ρ)
- transE isEquiv i j ρ = transS Den (i ρ) (j ρ)
+
+ isEquiv = record { refl  =         λ ρ → reflS  Domain
```

Comment:
this repeated code makes me think you're building a bunch of `IndexedSetoid`s "by hand".

Reply:
...yes, probably.  Feel free to refactor once I improve and merge this PR.